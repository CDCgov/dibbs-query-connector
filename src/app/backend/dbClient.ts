import { PoolConfig, Pool } from "pg";

// Load environment variables from .env and establish a Pool configuration
const dbConfig: PoolConfig = {
  connectionString: process.env.DATABASE_URL,
  max: 10, // Maximum # of connections in the pool
  idleTimeoutMillis: 30000, // A client must sit idle this long before being released
  connectionTimeoutMillis: process.env.LOCAL_DB_CLIENT_TIMEOUT
    ? Number(process.env.LOCAL_DB_CLIENT_TIMEOUT)
    : 3000, // Wait this long before timing out when connecting new client
};

let cachedDbClient: Pool | null = null;

/**
 * Getter function to retrieve the DB client from a naive cache and create a new one
 * if one doesn't exist
 * @returns a cached version of the DB client
 */
export const getDbClient = () => {
  if (!cachedDbClient) {
    cachedDbClient = new Pool(dbConfig);
  }
  return cachedDbClient;
};
