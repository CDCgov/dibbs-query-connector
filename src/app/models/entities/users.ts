import { CustomUserQuery } from "./query";

export enum UserRole {
  ADMIN = "Admin",
  STANDARD = "Standard",
  SUPER_ADMIN = "Super Admin",
}
export interface User {
  id: string;
  username: string;
  firstName: string;
  lastName: string;
  qcRole: UserRole;
  userGroupMemberships?: UserGroupMembership[];
}

export interface UserGroup {
  id: string;
  name: string;
  memberSize: number;
  querySize: number;
  members?: User[];
  queries?: CustomUserQuery[];
}

export interface UserGroupMembership {
  membershipId: string;
  usergroupName: string;
  usergroupId: string;
  isMember: boolean;
}

export interface Query {
  id: string;
  name: string;
  userGroupMemberships?: UserGroupMembership[];
}
