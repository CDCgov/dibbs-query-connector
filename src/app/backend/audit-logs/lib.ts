import { auth } from "@/auth";
import crypto from "crypto";

// these functions are exported from another file into the auditable file
//  to allow us to spy on them when calling @auditable functions
// Relevant thread: https://stackoverflow.com/questions/61111423/expected-spyon-function-to-be-called-jest

/**
 * Util function generate the success string after an audit write
 * @param actionType - the type of the audit information
 * @param id - id of the audit row
 * @param  checksum - the checksum generated by the audit contents
 * @returns Relevant info to track down the audit success message
 */
export function generateAuditSuccessMessage(
  actionType: string,
  id: string,
  checksum: string,
) {
  return `${actionType} audit action with id ${id} and checksum ${checksum} added to audit table`;
}

/**
 * Helper function that formats method input into the values needed by the audit
 * log table
 * @param methodName - name of the method to audit
 * @param argLabels - name of the params
 * @param args - value of the arguments
 * @returns an array of values to be inserted into the table
 */
export async function generateAuditValues(
  methodName: string,
  argLabels: string[],
  args: unknown[],
) {
  const session = await auth();
  const author = session?.user?.username || "No session information";
  const timestamp = new Date().toISOString();
  const auditContents = generateAuditMessage(argLabels, args);
  const auditChecksum = generateAuditChecksum(author, auditContents, timestamp);
  return [author, methodName, auditContents, timestamp, auditChecksum];
}

/**
 * Helper function that generates message contents
 * @param argLabels - the name of the parameters being passed into the function
 * @param args - the values of the parameters being passed in
 * @returns A JSON object with {argName: argValue}
 */
function generateAuditMessage(argLabels: string[], args: unknown[]) {
  // can do more fancy serialization here if needed
  const mappedArgs: { [s: string]: string } = {};
  args.forEach((obj, i) => {
    // Include undefined as null to ensure all arguments are captured
    if (obj !== undefined) {
      const val = JSON.stringify(obj);
      if (val !== "{}" && val !== "[]") {
        mappedArgs[argLabels[i]] = sanitizeString(JSON.stringify(obj));
      }
    } else {
      // Include undefined values as null in the audit log
      mappedArgs[argLabels[i]] = "null";
    }
  });

  return mappedArgs;
}

/**
 * Helper function that generates a checksum for audit contents based on audit
 * contents, author, and timestamp
 * @param author - name of the author
 * @param auditContents - the contents of the message being stored
 * @param timestamp - the time the message was created in the database
 * @returns A SHA-2 strength or greater checksum
 */
export function generateAuditChecksum(
  author: string,
  auditContents: object,
  timestamp: string,
) {
  const input = JSON.stringify({ author, auditContents, timestamp });
  return crypto.createHash("sha256").update(input).digest("hex");
}

function sanitizeString(s: string) {
  // add any other sanitization functions we need to this array
  const functionsToApply = [redactBearerToken];
  return functionsToApply.reduce((acc, cur) => cur(acc), s);
}

function redactBearerToken(s: string) {
  return s.replace(/"Bearer\s+([^\s]+)\"}/g, '"Bearer REDACTED"}');
}
