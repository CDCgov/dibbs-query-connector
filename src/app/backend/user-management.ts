"use server";

import { QueryResult } from "pg";
import {
  User,
  UserRole,
  UserGroup,
  UserGroupMembership,
} from "../models/entities/users";
import { QCResponse } from "../models/responses/collections";
import dbService from "./db/service";
import { adminRequired, superAdminRequired } from "./db/decorators";
import { auditable } from "./audit-logs/decorator";

class UserManagementService {
  /**
   * @param username The identifier of the user we want to retrieve
   * @returns A single user result, with any applicable group membership details
   */
  static async getUserByUsername(username: string): Promise<QCResponse<User>> {
    const userQuery = `SELECT * FROM users WHERE username = $1;`;
    const result = await dbService.query(userQuery, [username]);

    if (result.rowCount && result.rowCount > 0) {
      const user = result.rows[0];
      const userWithGroups =
        await UserManagementService.getSingleUserWithGroupMemberships(user.id);
      return {
        totalItems: userWithGroups.totalItems,
        items: userWithGroups.items,
      };
    } else {
      return { totalItems: 0, items: [] };
    }
  }

  /**
   * Adds a user to the users table if they do not already exist.
   * Uses data extracted from the JWT token.
   * @param userToken - The user data from the JWT token.
   * @param userToken.id - The user ID from the JWT token.
   * @param userToken.username - The username from the JWT token.
   * @param userToken.email - The email from the JWT token.
   * @param userToken.firstName - The first name from the JWT token.
   * @param userToken.lastName - The last name from the JWT token.
   * @returns The newly added user or an empty result if already exists.
   */
  @auditable
  static async addUserIfNotExists(userToken: {
    id: string;
    username: string;
    email: string;
    firstName: string;
    lastName: string;
  }) {
    if (!userToken || !userToken.username) {
      console.error("Invalid user token. Cannot add user.");
      return { user: undefined };
    }

    const { username, email, firstName, lastName } = userToken;
    const userIdentifier = username || email;

    try {
      console.log("Checking if user exists.");
      const checkUserQuery = `SELECT id, username, qc_Role, first_name, last_name FROM users WHERE username = $1;`;
      const userExists = await dbService.query(checkUserQuery, [
        userIdentifier,
      ]);

      if (userExists.rows.length > 0) {
        console.log("User already exists in users:", userExists.rows[0].id);
        return { msg: "User already exists", user: userExists.rows[0] }; // Return existing user
      }

      // Default role when adding a new user, which includes Super Admin, Admin, and Standard User.
      let qcRole = UserRole.STANDARD;
      console.log("User not found. Proceeding to insert.");

      const insertUserQuery = `
      INSERT INTO users (username, qc_role, first_name, last_name)
      VALUES ($1, $2, $3, $4)
      RETURNING id, username, qc_role, first_name, last_name;
    `;

      const result = await dbService.query(insertUserQuery, [
        userIdentifier,
        qcRole,
        firstName,
        lastName,
      ]);

      console.log("User added to users:", result.rows[0].id);
      return { user: result.rows[0] };
    } catch (error) {
      console.error("Error adding user to users:", error);
      throw error;
    }
  }

  /**
   * Updates an existing user in the users table.
   * @param userId - The user ID from the JWT token.
   * @param updated_userName - The username from the JWT token.
   * @param updated_firstName - The first name from the JWT token.
   * @param updated_lastName - The last name from the JWT token.
   * @param updated_role - The role from the JWT token.
   * @returns The newly added user or an empty result if already exists.
   */
  @auditable
  static async updateUserDetails(
    userId: string,
    updated_userName: string,
    updated_firstName: string,
    updated_lastName: string,
    updated_role: string,
  ) {
    if (!userId) {
      console.error("Cannot update user, no ID provided.");
      return;
    }

    try {
      console.log("Checking if user exists.");

      const checkUserQuery = `SELECT id, username, qc_role FROM users WHERE id = $1;`;
      const userExists = await dbService.query(checkUserQuery, [userId]);

      if (userExists.rowCount == 0) {
        console.log("User not found in Users", userId);
        return { msg: "Unable to update user", userId: userId };
      }

      const { username, qcRole, firstName, lastName } = userExists.rows[0];
      if (
        username !== updated_userName ||
        qcRole !== updated_role ||
        firstName !== updated_firstName ||
        lastName !== updated_lastName
      ) {
        const insertUserQuery = `
        UPDATE users
        SET 
          username = $2,
          first_name = $3,
          last_name = $4,
          qc_role = $5
        WHERE id = $1
        RETURNING id, username, qc_role, first_name, last_name;
      `;

        const result = await dbService.query(insertUserQuery, [
          userId,
          updated_userName,
          updated_firstName,
          updated_lastName,
          updated_role,
        ]);

        console.log("User updated:", result.rows[0].id);
        return result.rows[0];
      } else {
        // nothing to update, carry on...
        return;
      }
    } catch (error) {
      console.error("Error updating user", error);
      throw error;
    }
  }

  /**
   * Retrieves group membership data for the given users and formats it on the User object
   * @param userList - The users whose group memberships we are retrieving
   * @returns The updated user record or an error if the update fails.
   */
  static async getUserGroupMembershipDetails(userList: QueryResult) {
    const users = await Promise.all(
      userList.rows.map(async (user) => {
        try {
          const userGroups = await getAllUserGroupsForUser(user.id);
          user.userGroupMemberships = userGroups.items;
          return user;
        } catch (error) {
          console.error(error);
          throw error;
        }
      }),
    );

    return users;
  }

  /**
   * Updates the role of an existing user in the users table.
   * @param userId - The user ID.
   * @param newRole - The new role to assign to the user.
   * @returns The updated user record or an error if the update fails.
   */
  @superAdminRequired
  @auditable
  static async updateUserRole(
    userId: string,
    newRole: UserRole,
  ): Promise<QCResponse<User>> {
    try {
      console.log(`Updating role for user ID: ${userId} to ${newRole}`);

      const updateQuery = `
      UPDATE users
      SET qc_role = $1
      WHERE id = $2
      RETURNING id, username, qc_role, first_name, last_name;
    `;

      const result = await dbService.query(updateQuery, [newRole, userId]);

      if (result.rows.length === 0) {
        console.error(`User not found: ${userId}`);
        throw new Error(`User not found: ${userId}`);
      }

      console.log(`User role updated successfully: ${userId} -> ${newRole}`);
      const userWithGroups = await getUserGroupMembershipDetails(result);

      return { totalItems: 1, items: userWithGroups };
    } catch (error) {
      console.error("Error updating user role:", error);
      throw error;
    }
  }

  /**
   * Checks whether a given user exists
   * @param userId - The unique identifier for the user
   * @returns The given user, if it exists
   */
  static async checkUserExists(userId: string): Promise<QCResponse<User>> {
    if (!userId) {
      return { totalItems: 0, items: [] };
    }

    const userCheckQuery = `SELECT * FROM users WHERE id = $1;`;
    const userCheckResult = await dbService.query(userCheckQuery, [userId]);

    return {
      totalItems: userCheckResult.rowCount,
      items: userCheckResult.rows,
    } as QCResponse<User>;
  }

  /**
   * Retrieves all registered users in query connector
   * @returns List of users registered in qc
   */
  @adminRequired
  static async getAllUsers(): Promise<QCResponse<User>> {
    try {
      const selectAllUsersQuery = `
      SELECT id, username, qc_role, first_name, last_name
      FROM users
      ORDER BY last_name, first_name ASC;
    `;

      const result = await dbService.query(selectAllUsersQuery);
      const usersWithGroups = await getUserGroupMembershipDetails(result);

      return {
        totalItems: result.rowCount,
        items: usersWithGroups,
      } as QCResponse<User>;
    } catch (error) {
      throw error;
    }
  }

  /**
   * Retrieves the user group(s) for a given user
   * @param userId identifier of the user whose groups we are retrieving
   * @returns A list of UserGroup items for the given user
   */
  @adminRequired
  static async getAllUserGroupsForUser(
    userId: string,
  ): Promise<QCResponse<UserGroup[]>> {
    try {
      const selectAllUsersGroupsQuery = `
     SELECT  u.id as user_id, ug.name as usergroup_name, ug.id as usergroup_id, 
      COALESCE(member_count, 0) AS member_size,
      COALESCE(query_count, 0) AS query_size
    FROM usergroup_to_users as ugtu
    LEFT JOIN users as u ON u.id = ugtu.user_id  
    LEFT JOIN usergroup as ug ON ug.id = ugtu.usergroup_id
    LEFT JOIN (
      SELECT usergroup_id, COUNT(*) AS member_count 
      FROM usergroup_to_users 
      GROUP BY usergroup_id
    ) uu ON ug.id = uu.usergroup_id
    LEFT JOIN (
      SELECT usergroup_id, COUNT(*) AS query_count 
      FROM usergroup_to_query 
      GROUP BY usergroup_id
    ) uq ON ug.id = uq.usergroup_id
    WHERE ugtu.user_id = $1
    `;

      const result = await dbService.query(selectAllUsersGroupsQuery, [userId]);

      return {
        totalItems: result.rowCount,
        items: result.rows,
      } as QCResponse<UserGroup[]>;
    } catch (error) {
      console.error("Error retrieving user groups:", error);
      throw error;
    }
  }

  /**
   * Retrieves a user's role
   * @param username user's username. Username must be unique.
   * @returns The user's role or empty if the user is not found
   */
  static async getUserRole(username: string): Promise<string> {
    try {
      const selectUserRoleQuery = `
      SELECT id, username, qc_role, first_name, last_name
      FROM users
      WHERE username = $1;
    `;

      const result = await dbService.query(selectUserRoleQuery, [username]);
      return result?.rowCount && result.rowCount > 0
        ? result.rows?.[0].qcRole
        : "";
    } catch (error) {
      throw error;
    }
  }

  /**
   * Retrieves the full user object with all group membership(s)
   * @param userId - The unique identifier of the user.
   * @returns A User with an updated list of UserGroupMembership items
   */
  static async getSingleUserWithGroupMemberships(
    userId: string,
  ): Promise<QCResponse<User>> {
    const userCheckResult = await UserManagementService.checkUserExists(userId);
    if (!userCheckResult || userCheckResult?.totalItems == 0) {
      return { totalItems: 0, items: [] };
    }

    const query = `
      SELECT  ug.id as membership_id, g.id as usergroup_id, g.name as usergroup_name
      FROM usergroup_to_users as ug
      LEFT JOIN users as u
        ON u.id = ug.user_id
      LEFT JOIN usergroup as g
        ON g.id = ug.usergroup_id
      WHERE ug.user_id = $1;
    `;

    const result = await dbService.query(query, [userId]);

    const memberships: UserGroupMembership[] = result.rows.map((row) => {
      const { membershipId, usergroupId, usergroupName } = row;
      return {
        membershipId,
        usergroupId,
        usergroupName,
        isMember: true,
      };
    });

    const userWithGroups = {
      id: userId,
      username: userCheckResult.items[0].username,
      firstName: userCheckResult.items[0].firstName,
      lastName: userCheckResult.items[0].lastName,
      qcRole: userCheckResult.items[0].qcRole,
      userGroupMemberships: memberships,
    };

    return { totalItems: 1, items: [userWithGroups] };
  }

  /**
   * Retrieves the user group membership for a specific user group.
   * @param groupId - The unique identifier of the user group.
   * @returns A list of users with membership status for the specified group.
   */
  @adminRequired
  static async getAllUsersWithSingleGroupStatus(
    groupId: string,
  ): Promise<User[]> {
    const groupCheckQuery = `SELECT id FROM usergroup WHERE id = $1;`;
    const groupCheckResult = await dbService.query(groupCheckQuery, [groupId]);
    if (groupCheckResult.rows.length === 0) {
      return [];
    }

    const query = `
      SELECT u.id, u.username, u.first_name, u.last_name, u.qc_role, g.name AS group_name, g.id as usergroup_id,
        COALESCE(ug.id, '') AS membership_id,
        CASE 
          WHEN ug.user_id IS NOT NULL THEN TRUE
          ELSE FALSE
        END AS is_member
      FROM users u
      LEFT JOIN usergroup_to_users ug 
        ON u.id = ug.user_id AND ug.usergroup_id = $1
      LEFT JOIN usergroup g 
        ON ug.usergroup_id = g.id
        ORDER BY u.last_name, u.first_name;
    `;
    const result = await dbService.query(query, [groupId]);

    return result.rows.map((row) => ({
      id: row.id,
      username: row.username,
      firstName: row.firstName,
      lastName: row.lastName,
      qcRole: row.qcRole,
      userGroupMemberships: [
        {
          membershipId: row.membershipId,
          usergroupId: groupId,
          usergroupName: row.groupName,
          isMember: row.isMember,
        },
      ],
    }));
  }
}

export const getUserByUsername = UserManagementService.getUserByUsername;
export const addUserIfNotExists = UserManagementService.addUserIfNotExists;
export const updateUserDetails = UserManagementService.updateUserDetails;
export const updateUserRole = UserManagementService.updateUserRole;
export const getUserGroupMembershipDetails =
  UserManagementService.getUserGroupMembershipDetails;
export const getAllUserGroupsForUser =
  UserManagementService.getAllUserGroupsForUser;
export const getUserRole = UserManagementService.getUserRole;
export const getAllUsers = UserManagementService.getAllUsers;
export const getAllUsersWithSingleGroupStatus =
  UserManagementService.getAllUsersWithSingleGroupStatus;
export const getSingleUserWithGroupMemberships =
  UserManagementService.getSingleUserWithGroupMemberships;
