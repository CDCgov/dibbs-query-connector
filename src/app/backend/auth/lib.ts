import { Account, Profile } from "next-auth";
import { UserToken } from "../user-management";
import { JWT } from "@auth/core/jwt";
import { decodeJwt } from "jose";
import { UserRole } from "@/app/models/entities/users";
import { isAuthDisabledServerCheck } from "@/app/utils/auth";

export interface AuthStrategy {
  parseIdpResponseForUserToken(
    token: JWT,
    account: Account,
    profile: Profile,
  ): UserToken;
}

const ROLE_TO_ENUM_MAP: Record<string, UserRole> = {
  standard: UserRole.STANDARD,
  "super-admin": UserRole.SUPER_ADMIN,
  admin: UserRole.ADMIN,
};

export class AuthContext {
  private authStrategy: AuthStrategy;

  constructor(authStrategy: AuthStrategy) {
    this.authStrategy = authStrategy;
  }

  public setStrategy(authStrategy: AuthStrategy) {
    this.authStrategy = authStrategy;
  }

  public extendTokenWithExpirationTime(token: JWT) {
    const now = Math.floor(Date.now() / 1000);

    if (token.exp) {
      token.expiresIn = token.exp - now;
    }

    // handle expired tokens
    if (token.expiresIn && token.expiresIn <= 0) {
      return null;
    }

    return token;
  }

  public extendTokenWithUserInfo(token: JWT, userToken: UserToken) {
    if (isAuthDisabledServerCheck()) {
      token.role = UserRole.SUPER_ADMIN;
    } else if (token.role) {
      token.role = userToken.role;
    }
    token = { ...token, ...userToken };
    return token;
  }

  public parseIdpResponseForUserToken(
    token: JWT,
    account: Account,
    profile: Profile,
  ): UserToken {
    return this.authStrategy.parseIdpResponseForUserToken(
      token,
      account,
      profile,
    );
  }
}

const CLIENT_ID_NAME = process.env.AUTH_CLIENT_ID as string;
export class KeycloakAuthStrategy implements AuthStrategy {
  public parseIdpResponseForUserToken(
    token: JWT,
    account: Account,
    profile: Profile,
  ) {
    let role: UserRole = UserRole.STANDARD;

    if (account?.access_token) {
      let decodedToken = decodeJwt(account?.access_token);

      const keycloakRoles = decodedToken?.resource_access as {
        [CLIENT_ID_NAME]: {
          ["roles"]: string[];
        };
      };
      const userRoles = keycloakRoles && keycloakRoles[CLIENT_ID_NAME]?.roles;

      if (userRoles && userRoles[0] && ROLE_TO_ENUM_MAP[userRoles[0]]) {
        role = ROLE_TO_ENUM_MAP[userRoles[0]];
      } else {
        console.error(
          "No role found in Keycloak assignments. Falling back to standard access",
        );
      }
    } else {
      console.error(
        "No ID token found in Keycloak response. Falling back to standard access",
      );
    }

    const userToken = {
      id: profile.sub || "",
      username: profile.preferred_username || "",
      email: profile.email || "",
      firstName: profile.given_name || "",
      lastName: profile.family_name || "",
      role: role,
    };
    return userToken;
  }
}

export class MicrosoftEntraAuthStrategy implements AuthStrategy {
  public parseIdpResponseForUserToken(
    token: JWT,
    account: Account,
    profile: Profile,
  ) {
    const azureRoles = profile?.roles as string[];
    const userInfo = decodeJwt(account?.id_token as string);
    let role = ROLE_TO_ENUM_MAP[azureRoles[0]];

    const fullName = userInfo.name as string;
    const firstName = token.given_name ?? fullName.split(" ")[0];
    const lastName = token.last_name ?? fullName.split(" ").slice(1).join(" ");

    const userToken = {
      id: profile.oid as string,
      username: profile?.preferred_username as string,
      firstName: firstName as string,
      lastName: lastName as string,
      role: role,
    };

    return userToken;
  }
}
