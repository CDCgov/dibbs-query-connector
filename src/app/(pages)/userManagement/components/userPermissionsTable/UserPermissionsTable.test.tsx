import { waitFor, screen, render } from "@testing-library/react";
import * as UserManagementBackend from "@/app/backend/user-management";
import { UserRole } from "@/app/models/entities/users";
import { renderWithUser, RootProviderMock } from "@/app/tests/unit/setup";
import UserPermissionsTable from "./userPermissionsTable";
import {
  mockAdmin,
  mockSuperAdmin,
  mockPermissionsTab,
} from "../../test-utils";
import { createRef, RefObject } from "react";

jest.mock("next-auth/react");

jest.mock(
  "@/app/ui/components/withAuth/WithAuth",
  () =>
    ({ children }: React.PropsWithChildren) => <div>{children}</div>,
);

jest.mock("@/app/backend/user-management", () => ({
  getAllUsers: jest.fn().mockResolvedValue({ items: [], totalItems: 0 }),
  getUserGroups: jest.fn().mockResolvedValue({ items: [], totalItems: 0 }),
  updateUserRole: jest.fn(),
}));

jest.mock("@/app/backend/query-building/service", () => ({
  getConditionsData: jest.fn(),
}));

const mockRefs = createRef<RefObject<HTMLTableRowElement | null>[]>();
mockRefs.current = [];

describe("User Management: User tab", () => {
  it("Loads user list successfully", async () => {
    jest.spyOn(UserManagementBackend, "getAllUsers").mockResolvedValue({
      items: [mockAdmin, mockSuperAdmin],
      totalItems: 2,
    });

    render(
      <RootProviderMock currentPage="/userManagement">
        <UserPermissionsTable
          openModal={jest.fn()}
          setUsers={jest.fn()}
          users={[mockAdmin, mockSuperAdmin]}
          fetchGroupMembers={jest.fn()}
          rowFocusRefs={mockRefs}
          modalData={""}
        />
      </RootProviderMock>,
    );
    await waitFor(() => {
      expect(screen.queryByText("Loading")).not.toBeInTheDocument();
    });

    if (mockPermissionsTab.renderContent) {
      render(mockPermissionsTab.renderContent());
    }
    expect(screen.queryAllByText("Potter, Harry")).toBeTruthy();
    expect(screen.queryAllByText("Potter, Lily")).toBeTruthy();
    expect(document.body).toMatchSnapshot();
  });

  it("Changes a user's role", async () => {
    jest.spyOn(UserManagementBackend, "getAllUsers").mockResolvedValue({
      items: [mockAdmin, mockSuperAdmin],
      totalItems: 2,
    });

    const updateRoleFnSpy = jest
      .spyOn(UserManagementBackend, "updateUserRole")
      .mockResolvedValue({
        totalItems: 1,
        items: [
          {
            ...mockAdmin,
            qcRole: UserRole.SUPER_ADMIN,
          },
        ],
      });

    const { user } = renderWithUser(
      <RootProviderMock currentPage="/userManagement">
        <UserPermissionsTable
          openModal={jest.fn()}
          setUsers={jest.fn()}
          users={[mockAdmin, mockSuperAdmin]}
          fetchGroupMembers={jest.fn()}
          rowFocusRefs={mockRefs}
          modalData={""}
        />
      </RootProviderMock>,
    );

    await waitFor(() => {
      expect(screen.queryByText("Loading")).not.toBeInTheDocument();
    });

    if (mockPermissionsTab.renderContent) {
      render(mockPermissionsTab.renderContent());
    }

    const adminDropdown = screen.queryAllByTestId(
      `role-select-${mockAdmin.id}`,
    )[0];

    expect(adminDropdown).toHaveValue("Admin");
    await user.selectOptions(adminDropdown, ["Super Admin"]);
    expect(adminDropdown).toHaveValue("Super Admin");
    expect(updateRoleFnSpy).toHaveBeenCalledWith(mockAdmin.id, "Super Admin");
  });
});
