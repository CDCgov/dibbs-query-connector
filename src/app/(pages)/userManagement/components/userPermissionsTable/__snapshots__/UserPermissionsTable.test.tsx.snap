// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Management: User tab Loads user list successfully 1`] = `
<body>
  <div>
    <div
      class="userPermissionsTable customizeTableWrapper"
    >
      <table
        class="usa-table usa-table--borderless customizeTable"
        data-testid="table"
      >
        <thead>
          <tr>
            <th>
              Name
            </th>
            <th>
              Permissions
            </th>
            <th>
              User groups
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="tableBody__row"
            id="456"
            tabindex="0"
          >
            <td>
              <div
                class="userDisplay"
              >
                <span>
                  Potter, Lily 
                </span>
                <button
                  class="usa-button usa-button--unstyled editBtn unstyled-button-container usa-button--unstyled text-bold text-no-underline"
                  data-testid="button"
                  type="button"
                >
                  <span
                    class="icon-text padding-right-4 display-flex flex-align-center"
                  >
                    <svg
                      aria-label="Pencil icon indicating edit ability"
                      class="usa-icon height-3 width-3"
                      focusable="false"
                      height="1em"
                      role="img"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 0 0 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                      />
                    </svg>
                    <span
                      class="padding-left-05"
                      data-testid="edit-group-456"
                    >
                      Edit
                    </span>
                  </span>
                </button>
              </div>
            </td>
            <td
              class="roleDropdown"
            >
              <label
                class="usa-label usa-sr-only"
                data-testid="label"
                for="role-select-456"
              >
                User role
              </label>
              <select
                class="usa-select"
                data-testid="role-select-456"
                id="role-select-456"
              >
                <option
                  value="Super Admin"
                >
                  Super Admin
                </option>
                <option
                  selected=""
                  value="Admin"
                >
                  Admin
                </option>
                <option
                  value="Standard"
                >
                  Standard
                </option>
              </select>
            </td>
            <td>
              <button
                aria-description="Edit Order of the Phoenix members"
                class="usa-button usa-button--unstyled margin-right-2 text-no-underline"
                data-testid="button"
                type="button"
              >
                <span>
                  Order of the Phoenix
                </span>
              </button>
            </td>
          </tr>
          <tr
            class="tableBody__row"
            id="123"
            tabindex="0"
          >
            <td>
              <div
                class="userDisplay"
              >
                <span>
                  Potter, Harry 
                </span>
                <button
                  class="usa-button usa-button--unstyled editBtn unstyled-button-container usa-button--unstyled text-bold text-no-underline"
                  data-testid="button"
                  type="button"
                >
                  <span
                    class="icon-text padding-right-4 display-flex flex-align-center"
                  >
                    <svg
                      aria-label="Pencil icon indicating edit ability"
                      class="usa-icon height-3 width-3"
                      focusable="false"
                      height="1em"
                      role="img"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 0 0 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                      />
                    </svg>
                    <span
                      class="padding-left-05"
                      data-testid="edit-group-123"
                    >
                      Edit
                    </span>
                  </span>
                </button>
              </div>
            </td>
            <td
              class="roleDropdown"
            >
              <label
                class="usa-label usa-sr-only"
                data-testid="label"
                for="role-select-123"
              >
                User role
              </label>
              <select
                class="usa-select"
                data-testid="role-select-123"
                id="role-select-123"
              >
                <option
                  selected=""
                  value="Super Admin"
                >
                  Super Admin
                </option>
                <option
                  value="Admin"
                >
                  Admin
                </option>
                <option
                  value="Standard"
                >
                  Standard
                </option>
              </select>
            </td>
            <td>
              --
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div>
    <div
      class="userPermissionsTable customizeTableWrapper"
    >
      <table
        class="usa-table usa-table--borderless customizeTable"
        data-testid="table"
      >
        <thead>
          <tr>
            <th>
              Name
            </th>
            <th>
              Permissions
            </th>
            <th>
              User groups
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="tableBody__row"
            id="456"
            tabindex="0"
          >
            <td>
              <div
                class="userDisplay"
              >
                <span>
                  Potter, Lily 
                </span>
                <button
                  class="usa-button usa-button--unstyled editBtn unstyled-button-container usa-button--unstyled text-bold text-no-underline"
                  data-testid="button"
                  type="button"
                >
                  <span
                    class="icon-text padding-right-4 display-flex flex-align-center"
                  >
                    <svg
                      aria-label="Pencil icon indicating edit ability"
                      class="usa-icon height-3 width-3"
                      focusable="false"
                      height="1em"
                      role="img"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 0 0 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                      />
                    </svg>
                    <span
                      class="padding-left-05"
                      data-testid="edit-group-456"
                    >
                      Edit
                    </span>
                  </span>
                </button>
              </div>
            </td>
            <td
              class="roleDropdown"
            >
              <label
                class="usa-label usa-sr-only"
                data-testid="label"
                for="role-select-456"
              >
                User role
              </label>
              <select
                class="usa-select"
                data-testid="role-select-456"
                id="role-select-456"
              >
                <option
                  value="Super Admin"
                >
                  Super Admin
                </option>
                <option
                  selected=""
                  value="Admin"
                >
                  Admin
                </option>
                <option
                  value="Standard"
                >
                  Standard
                </option>
              </select>
            </td>
            <td>
              <button
                aria-description="Edit Order of the Phoenix members"
                class="usa-button usa-button--unstyled margin-right-2 text-no-underline"
                data-testid="button"
                type="button"
              >
                <span>
                  Order of the Phoenix
                </span>
              </button>
            </td>
          </tr>
          <tr
            class="tableBody__row"
            id="123"
            tabindex="0"
          >
            <td>
              <div
                class="userDisplay"
              >
                <span>
                  Potter, Harry 
                </span>
                <button
                  class="usa-button usa-button--unstyled editBtn unstyled-button-container usa-button--unstyled text-bold text-no-underline"
                  data-testid="button"
                  type="button"
                >
                  <span
                    class="icon-text padding-right-4 display-flex flex-align-center"
                  >
                    <svg
                      aria-label="Pencil icon indicating edit ability"
                      class="usa-icon height-3 width-3"
                      focusable="false"
                      height="1em"
                      role="img"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 0 0 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                      />
                    </svg>
                    <span
                      class="padding-left-05"
                      data-testid="edit-group-123"
                    >
                      Edit
                    </span>
                  </span>
                </button>
              </div>
            </td>
            <td
              class="roleDropdown"
            >
              <label
                class="usa-label usa-sr-only"
                data-testid="label"
                for="role-select-123"
              >
                User role
              </label>
              <select
                class="usa-select"
                data-testid="role-select-123"
                id="role-select-123"
              >
                <option
                  selected=""
                  value="Super Admin"
                >
                  Super Admin
                </option>
                <option
                  value="Admin"
                >
                  Admin
                </option>
                <option
                  value="Standard"
                >
                  Standard
                </option>
              </select>
            </td>
            <td>
              --
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</body>
`;
